### ISSUE TITLE
Allow applications to process transactions in batch

### ISSUE THREAD

**Issue by danil-lashin** [tendermint/tendermint#2901]
### Problem

Developers cannot optimize transaction processing in applications because they receive one tx per time.

For example, the most expensive part of tx execution is public key derivation (ECDSA) from tx's signature. Derivation can be done in parallel and this approach will give us several times speed-up. 

### Solution

Allow applications to process transactions in batch.

I've started some work here: https://github.com/danil-lashin/tendermint/commit/370a0850a0845b94f3f1a8f35dee04a2791abbfd. If you agree with this approach, I will be happy to make a complete PR.

**Comment by ebuchman** [comment]
DeliverTx is sent asynchronously, but the responses are expected to come back in order. Thus it should be possible on the application side to process the txs concurrently so long as responses are sent back in the right order. Of course it means txs need to be buffered on the application side.

That said, for in-process apps, the DeliverTx calls are actually made synchronously. We should consider an asynchronous `localClient` to facilitate this.

Would that be sufficient? It should be possible to test this already by running your application out-of-process and using the standard Golang abci server

**Comment by ebuchman** [comment]
Another option would be to just send all the block data as you've suggested. We could consider doing this, but it would likely need to be configurable (ie. on startup apps will decide if they want individual txs or all of them at once).

**Comment by ValarDragon** [comment]
> DeliverTx is sent asynchronously, but the responses are expected to come back in order. Thus it should be possible on the application side to process the txs concurrently so long as responses are sent back in the right order. Of course it means txs need to be buffered on the application side.

Can this really be used for effective parallelism? You still likely want guarantees on the tx ordering, you just want to execute some part in parallel, and it may be a non-commutative relation. (So it can't be asynchronous / out of order) Signature verification is still commutative so I see how that could be done kinda hackily still, but waiting asynchronously for more to be delivered is kind of annoying / overall likely less efficient. (As you'd stall if you were doing batching + parallelism as you were waiting to build your batch)

**Comment by ebuchman** [comment]
Not sure I understand - ABCI delivery is asynchronous but in order.

**Comment by ValarDragon** [comment]
I'm confused as to how that is asynchronous then. Does it just wait if it gets things out of order?

**Comment by srmo** [comment]
AFAIUI, it just pumps deliverTx in order but async wrt not waiting for a response before sending the next deliverTx. 
Only after sending all txs it waits for all responses.

**Comment by ebuchman** [comment]
> AFAIUI, it just pumps deliverTx in order but async wrt not waiting for a response before sending the next deliverTx.
> Only after sending all txs it waits for all responses.

Correct.

It's asynchronous because we do not wait for a response from one DeliverTx before sending the next.

It's ordered because we throw an error if responses come back in the wrong error.

**Comment by ValarDragon** [comment]
Thanks for the clarification! I forgot that we ran TCP on our sockets, so we have in order delivery.

So what I was saying w.r.t. commutativity doesn't apply, but we do still have stalls for batched algorithms. I now think you're right that signature parallelization for what this issue is about can be done for the app when its not in process :)

I do think that sending all the txs at once is more optimal, but that discussion is probably best suited for a separate issue.

**Comment by ebuchman** [comment]
@danil-lashin what are your thoughts given the above? Should we still consider a DeliverTxs ABCI endpoint and an option for apps to get all the txs at once vs. one-by-one? If someone wants to write up an ADR for this, I've reserved [ADR-037](https://github.com/tendermint/tendermint/issues/2313) for this

**Comment by b00f** [comment]
@ebuchman @ValarDragon If you agree about introducing new ABCI method for `DeliverBlock(b Block)` I can make a PR for this change.

**Comment by jaekwon** [comment]
Here's why I think we should keep ABCI as is.

A Protobuf message is a binary blob.  I'm not aware of any protobuf clients that handle streaming w/ callbacks upon elements of an array, but that's essentially what you'll need if you want to be able to parse a tx and begin working on it the moment you receive it, rather than waiting for the whole set of txs to be received before processing.  Which means if the block size is large and there are many txs, or the txs are large, then block response time becomes *slower*, given the state of Protobuf clients and usage.  If we keep ABCI as is, then the only thing we need to do is optimize the Tendermint side in Go -- AFAIK there's no reason why streaming individual txs through ABCI would be significantly slower than sending a batch encoding, meanwhile the streaming of individual txs enables potentially signficant time savings in starting tx processing.

Also, this leaves the door open for more flexibility, such as filtering txs for shards of state machines.  A Tendermint node might be talking to multiple shards.  It might give each shard the same BeginBlock/EndBlock/Commit, but it might not give them all the same txs.  If you want to have a single DeliverBlock message, it's implying that you're going to construct new arrays in memory for each shard which may be large.  So I don't think we should assume that ABCI is deliverying whole blocks to the state machine.

And take a look at Commit... in the future we may want this message to happen more or less frequently.  Maybe users will want a Tendermint option to store intermediate app hashes.  Or maybe we'll want to split RequestCommit into RequestHash and RequestSave, and call RequestSave less frequently than every block.

It's easy enough to write shims that combines the Begin/Tx(s)/End/Commit into a single request/response object, so I think we should just keeps things as is, which implement the minimum subset of functionality for near optimum performance without needing to write intra-message-streaming protobuf decoders in all languages.

**Comment by ebuchman** [comment]
> there's no reason why streaming individual txs through ABCI would be significantly slower than sending a batch encoding, 

"no reason" seems like a strong statement here. The overhead is sending a serialized TCP message over a socket. The app could be running on another machine, in which case there will also probably be encryption overhead.

> meanwhile the streaming of individual txs enables potentially signficant time savings in starting tx processing.

Isn't this somewhat contradictory with the previous point? The previous point was that the overhead of sending individual ABCI msgs is negligible, but this point is that you can save "significant" time by not waiting for them.

I think the even bigger consideration (and the one mentioned in the OP) is the potential optimizations in signature verification if you can get the whole block at once, which will blow away any savings from being able to process msgs before the next one is received.

> So I don't think we should assume that ABCI is deliverying whole blocks to the state machine.

Seems like it should be configurable. It's not much additional complexity to support both options and will probably be warranted.

> It's easy enough to write shims that combines the Begin/Tx(s)/End/Commit into a single request/response object, so I think we should just keeps things as is, which implement the minimum subset of functionality for near optimum performance without needing to write intra-message-streaming protobuf decoders in all languages.

How does this work without forking Tendermint?

**Comment by jaekwon** [comment]
> "no reason" seems like a strong statement here. The overhead is sending a serialized TCP message over a socket. The app could be running on another machine, in which case there will also probably be encryption overhead.

That overhead was going to happen anyways, in a different form.

> Isn't this somewhat contradictory with the previous point? The previous point was that the overhead of sending individual ABCI msgs is negligible, but this point is that you can save "significant" time by not waiting for them.

No, one is about latency of how long it takes from creation of a block until it's committed w/ resulting app hash, and the other is the overall time.

A======B---------------------------C

A: block creation
B: start executing 1st tx on state machine
C: when Tendermint receives the commit app hash

The former is about A-B.  The latter is about A-C.  I'm saying A-B is potentially very significant for some fraction of apps, while any differences in A-C can be optimized in Tendermint such that it doesn't matter for any application.

> I think the even bigger consideration (and the one mentioned in the OP) is the potential optimizations in signature verification if you can get the whole block at once, which will blow away any savings from being able to process msgs before the next one is received.

But you can do that already with a shim, so nothing is added by changing ABCI, so there's no benefit to making changes, whereas some apps *won't* benefit from this optimization.

> Seems like it should be configurable. It's not much additional complexity to support both options and will probably be warranted.

I think we should support many features in Tendermint, but orthogonal ones that provide tangible benefits.

> How does this work without forking Tendermint?

I'm talking about future versions of Tendermint, so technically forks yes.  By keeping ABCI as is we can more easily support existing clients for the new version as well.

**Comment by thith** [comment]
FYI, in our app, we are actually trying to run txs in parallel. It is js-abci, out-of-process app.
1. With DeliverBlock: this makes it very simple for use cases like ours. In addition, it is a new abci option so it won't affect existing apps.
2. With DeliverTx: we will have to buffer to wait for all txs to arrive before deciding which ones could be run in parallel. I don't know how to do this with the existing js-abci, so maybe I will have to fork it. It's more complex. However, receiving first TX earlier to do some preprocess is a good point though.

As this proposal is undecided waiting for a more convincing use cases, for now, I will try with the DeliverTx option.

**Comment by b00f** [comment]
Running all Txs at once, gives more options to application to control block structure.
Running Txs one by one gives more power to tednermint for its future thoughts and designs.

Depends where are you staying, your vote will change. ;)
Maybe you need to do a survey for this change.
